java/com/google/location/lbs/activity/audioset/dataprocessing/CsvParser.java

```diff
@@ -0,0 +1,244 @@
+import java.io.File;
+import java.io.IOException;
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Arrays;
+
+/**
+ * CSV Class
+ *
+ * <p>CSV is a class for the reading of a dataset in a csv format.
+ *
+ * @author Carver Forbes
+ * @version 0.1
+ */
+public class CsvParser {
+
+  private File file;
+  private ArrayList<String> ytidList;
+  private HashMap<String, IncompleteExample.Example> idToProtoMap;
+  private HashMap<String, String> idToUrlMap;
+
+  private final boolean debug = true;
+
+  /**
+   * CSV constructor
+   *
+   * @param pathname the pathname for the csv file to be read
+   * @throws IllegalArgumentException if the pathname is incorrect
+   */
+  public CsvParser(String pathname) {
+    file = new File(pathname);
+    if (!file.exists()) {
+      throw new IllegalArgumentException(String.format("This file does not exist: %s", pathname));
+    }
+    ytidList = new ArrayList<>();
+    idToProtoMap = new HashMap<>();
+    idToUrlMap = new HashMap<>();
+  }
+
+  /** Tester Function to print 5 YouTube Links */
+  public void printYouTubeLinks() {
+    // prints youtube links
+    Scanner scan = null;
+    try {
+      scan = new Scanner(file);
+      String line = scan.nextLine();
+      while (line.charAt(0) == '#') {
+        line = scan.nextLine();
+      }
+      // while (scan.hasNext()) {
+      //   String ytid = line.split(",")[0];
+      //   String url = "https://youtube.com/watch?v=" + ytid;
+      //   System.out.println(url);
+      //   line = scan.nextLine();
+      // }
+      for (int i = 0; i < 5; i++) {
+        String ytid = line.split(",")[0];
+        String url = "https://youtube.com/watch?v=" + ytid;
+        System.out.println(url);
+        line = scan.nextLine();
+      }
+    } catch (IOException io) {
+      io.printStackTrace();
+    } finally {
+      if (scan != null) {
+        scan.close();
+      }
+    }
+  }
+
+  /** Extracts metadata from dataset.csv file */
+  public void csvToProtoMap() {
```

Comment:
looks like this method needs to run before any other action -- better to merge it with the constructor?

Reply:
Having a lot of work done in the constructor is generally a [code smell](https://martinfowler.com/bliki/CodeSmell.html#:~:text=A%20code%20smell%20is%20a,me%20with%20my%20Refactoring%20book.&text=The%20best%20smells%20are%20something,you%20to%20really%20interesting%20problems.) to me. It seems like what we actually want is a method, even a static one. Maybe refactor this class into a static final class named `AudioSetCsvParser` that contains one static method `Map<String, Example> parse(String filename)`?

Reply:
The csvToProtoMap() method is meant to be called by the main method. idToProtoMap is only supposed to be initialized to a an empty HashMap {video_id: protocol buffer} in the constructor.